src/librustc_typeck/impl_wf_check/min_specialization.rs

```diff
@@ -0,0 +1,315 @@
+
+        check_static_lifetimes(tcx, &parent_substs, span);
+        check_duplicate_params(tcx, impl1_substs, &parent_substs, span);
+
+        check_predicates(tcx, impl1_def_id, impl1_substs, impl2_node, impl2_substs, span);
+    }
+}
+
+// Get substs for the two impls so that they have the same `TraitRef`.
+fn get_impl_substs<'tcx>(
+    infcx: &InferCtxt<'_, 'tcx>,
+    impl1_def_id: DefId,
+    impl2_node: Node,
+    span: Span,
+) -> Option<(SubstsRef<'tcx>, SubstsRef<'tcx>)> {
+    let tcx = infcx.tcx;
+    let param_env = tcx.param_env(impl1_def_id);
+
+    let impl1_substs = InternalSubsts::identity_for_item(tcx, impl1_def_id);
+    let impl2_substs = translate_substs(infcx, param_env, impl1_def_id, impl1_substs, impl2_node);
+
+    // Conservatively use an empty `ParamEnv`.
+    let outlives_env = OutlivesEnvironment::new(ty::ParamEnv::empty());
+    infcx.resolve_regions_and_report_errors(
+        impl1_def_id,
+        &ScopeTree::default(),
+        &outlives_env,
+        SuppressRegionErrors::default(),
+    );
+    let impl2_substs = match infcx.fully_resolve(&impl2_substs) {
+        Ok(s) => s,
+        Err(_) => {
+            tcx.sess.struct_span_err(span, "could not resolve substs on overridden impl").emit();
+            return None;
+        }
+    };
+    Some((impl1_substs, impl2_substs))
+}
+
+/// Returns a list of all of the unconstrained subst of the given impl.
+fn unconstrained_parent_impl_substs<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl_def_id: DefId,
+    impl_substs: SubstsRef<'tcx>,
+) -> Vec<GenericArg<'tcx>> {
+    let mut seen_params = FxHashSet::default();
+    let mut constrained_params = FxHashSet::default();
+
+    let impl2_identity_substs = InternalSubsts::identity_for_item(tcx, impl_def_id);
+    let impl2_generic_predicates = tcx.predicates_of(impl_def_id);
+    let impl2_identity_predicates =
+        impl2_generic_predicates.instantiate(tcx, impl2_identity_substs);
+
+    for predicate in impl2_identity_predicates.predicates.iter() {
+        if let ty::Predicate::Projection(proj) = predicate {
+            let projection_ty = proj.skip_binder().projection_ty;
+            let projected_ty = proj.skip_binder().ty;
+            seen_params.extend(cgp::parameters_for(&projection_ty, true));
+            for param in cgp::parameters_for(&projected_ty, false) {
+                if !seen_params.contains(&param) {
+                    constrained_params.insert(param.0);
+                }
+            }
+            seen_params.extend(cgp::parameters_for(&projected_ty, true));
+        }
+    }
+
+    impl_substs
+        .iter()
+        .enumerate()
+        .filter(|(idx, _)| !constrained_params.contains(&(*idx as u32)))
+        .map(|(_, arg)| *arg)
+        .collect()
+}
+
+fn check_duplicate_params<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl1_substs: SubstsRef<'tcx>,
+    parent_substs: &Vec<GenericArg<'tcx>>,
+    span: Span,
+) {
+    let mut base_params = cgp::parameters_for(parent_substs, true);
```

Comment:
It would be good to comment why you chose this function. I think the example would be something like:

```rust
impl<A> Foo for A { }
impl<B> Foo for dyn Iterator<Item = (B, B)> { }
```

and here we want to say that, even though `B` repeats, it's ok because it's constrained?

Another example might be like `dyn Foo<B, Out = B>` -- this would also be approved.

But, thinking on it more, is that really correct? Like consider the type `dyn Foo<&'a u32, Out = &'b u32>`. How can we decide whether the impl applies? We'd have to know that `'a = 'b`.

Reply:
`parameters_for(_, true)` doesn't have any special handling of any `TyKind`. I'm using it to conveniently handle the 3 different kinds of params.

Reply:
Oh, I see. Man, I hate random booleans in the signature, but pre-existing I suppose. :)

Reply:
And quite possibly my fault.